include/power/power_state.h

```diff
@@ -103,6 +103,151 @@ enum pm_state {
 	PM_STATE_SOFT_OFF
 };
 
+/**
+ * Information about a power management state
+ */
+struct pm_state_info {
+	enum pm_state state;
+	int32_t min_residency;
+};
+
+/**
+ * @brief Construct a pm_state_info from 'pm_states' property at index 'i'
+ *
+ * @param node_id A node identifier compatible with pm-states
+ * @param i index of pm_states prop which type is 'phandles'
+ * @return pm_state_info item from 'pm_states' property at index 'i'
+ */
+#define DT_PM_STATE_INFO_ITEM_BY_IDX(node_id, i)                    \
+	{                                                           \
+		.state = DT_PROP_BY_PHANDLE_IDX(node_id,            \
+					pm_states, i, pm_state),    \
+		.min_residency = DT_PROP_BY_PHANDLE_IDX_OR(node_id, \
+				pm_states, i, min_residency, -1),   \
+	},
+
+/**
+ * @brief Length of 'pm-states' property which type is 'phandles'
+ *
+ * @param node_id A node identifier compatible with pm-states
+ * @return length of 'pm-states' property which type is 'phandles'
+ */
+#define DT_PM_STATE_ITEMS_LEN(node_id) DT_PROP_LEN(node_id, pm_states)
+
+/**
+ * @brief Macro function to construct enum pm_state item in UTIL_LISTIFY
+ * extension.
+ *
+ * @param child child index in UTIL_LISTIFY extension.
+ * @param node_id A node identifier compatible with pm-states
+ * @return macro function to construct a pm_state_info
+ */
+#define DT_PM_STATE_INFO_ITEMS_FUC(child, node_id) \
+	DT_PM_STATE_INFO_ITEM_BY_IDX(node_id, child)
+
+/**
+ * @brief Macro function to construct a list of 'pm_state_info' items by
+ * UTIL_LISTIFY func
+ *
+ * Example devicetree fragment:
+ *	cpus {
+ *		...
+ *		cpu0: cpu@0 {
+ *			device_type = "cpu";
+ *			...
+ *			pm-states = <&state0 &state1>;
```

Comment:
This is structurally very different from [Linux](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/devicetree/bindings/arm/idle-states.yaml#n479).  Does it need to be that different?

Reply:
```
  - |

            CPU_SLEEP_1_0: cpu-sleep-1-0 {
                compatible = "arm,idle-state";
                local-timer-stop;
                arm,psci-suspend-param = <0x0010000>;
                entry-latency-us = <70>;
                exit-latency-us = <100>;
                min-residency-us = <300>;
                wakeup-latency-us = <150>;
            };
...
    cpus {
        #size-cells = <0>;
        #address-cells = <1>;

        cpu@0 {
            device_type = "cpu";
            compatible = "arm,cortex-a15";
            reg = <0x0>;
            cpu-idle-states = <&cpu_sleep_1_0>;
        };
```

What exactly is that different ? Linux has more properties, some of them we will probably add later, I just didn't put now because we don't have use for then currently and it is always easy adding something than remove. Regarding names, I saw more power/pm than idle in our code.

But IMHO the struct is the same, we have power state declaration and and cpu has a property with a list of them ....

Reply:
OK, I see it's not as structurally different as I thought, it's mostly that the property and compatible names don't match.

We have generally tried to be compatible at devicetree source unless there's some clear reason we can't do that.  I'm not convinced we can as Linux has target-specific compatibles and perhaps we're aiming to be target-agnostic, but I do think we should match the property names that aren't ARM or other vendor specific, such as the latencies and timer effects.

Reply:
I tend to agree as well. In this pr, it translate to `s/min-residency/min-residency-us/`